packages/lib-utils/src/app/api-discovery/discovery-cache.ts

```diff
@@ -1,23 +1,51 @@
 import { consoleLogger } from '@monorepo/common';
+import * as _ from 'lodash-es';
+import { getReferenceForModel } from '../../k8s/k8s-utils';
 import type { DiscoveryResources } from '../../types/api-discovery';
+import type { K8sModelCommon } from '../../types/k8s';
 
 const SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY = 'sdk/api-discovery-resources';
 
-export const cacheResources = (resources: DiscoveryResources) => {
+const mergeByKey = (prev: K8sModelCommon[], next: K8sModelCommon[]) =>
+  Object.values(_.merge(_.keyBy(prev, getReferenceForModel), _.keyBy(next, getReferenceForModel)));
+
+export const cacheResources = (resources: DiscoveryResources[]) => {
+  let allResources;
+  try {
+    allResources = [
+      ...[JSON.parse(localStorage.getItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY) || '{}')],
+      ...resources,
+    ].reduce((acc, curr) => {
+      const { models, configResources, clusterOperatorConfigResources, ...rest } = curr || {};
+      return {
+        ..._.merge(acc, rest),
+        configResources: mergeByKey(acc?.configResources, configResources || []),
+        models: mergeByKey(acc?.models, models || []),
+        clusterOperatorConfigResources: mergeByKey(
+          acc?.clusterOperatorConfigResources,
+          clusterOperatorConfigResources || [],
+        ),
+      };
+    }, {});
+  } catch (e) {
+    consoleLogger.error('Error caching API resources in localStorage', e);
```

Comment:
I assume that only JSON parse & local storage access may result in an error, otherwise it's just data processing.

I'd suggest adding functions to handle the locally cached API resources, for example:

```ts
const getLocalResources = () => {
  try {
    return JSON.parse(localStorage.getItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY) || '{}');
  } catch (e) {
    throw new Error('Cannot load cached API resources', e);
  }
};

const setLocalResources = (resources) => {
  try {
    localStorage.setItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY, JSON.stringify(resources));
  } catch (e) {
    throw new Error('Cannot write cached API resources', e);
  }
};
```

Reply:
@vojtechszocs agreed

Reply:
Good point!